nuget/helpers/lib/NuGetUpdater/DotNetPackageCorrelation/Correlator.cs

```diff
@@ -0,0 +1,197 @@
+            var releasesFile = JsonSerializer.Deserialize<ReleasesFile>(releasesJson, SerializerOptions)!;
+
+            foreach (var release in releasesFile.Releases)
+            {
+                foreach (var sdk in release.GetSdks())
+                {
+                    if (sdk.Version is null)
+                    {
+                        warnings.Add($"Skipping release with missing version information from {releasesJson}");
+                        continue;
+                    }
+
+                    if (sdk.RuntimeVersion is null)
+                    {
+                        warnings.Add($"Skipping release with missing runtime version information from {releasesJson}");
+                        continue;
+                    }
+
+                    if (runtimePackages.Runtimes.ContainsKey(sdk.RuntimeVersion))
+                    {
+                        // already processed this runtime
+                        continue;
+                    }
+
+                    var packagesAndVersions = new PackageSet();
+                    runtimePackages.Runtimes.Add(sdk.RuntimeVersion, packagesAndVersions);
+                    var runtimeDirectory = new DirectoryInfo(Path.Combine(_releaseNotesDirectory.FullName, majorVersion.ToString(), sdk.RuntimeVersion.ToString()));
+                    var runtimeMarkdownPath = Path.Combine(runtimeDirectory.FullName, $"{sdk.RuntimeVersion}.md");
+                    if (!File.Exists(runtimeMarkdownPath))
+                    {
+                        warnings.Add($"Unable to find expected markdown file {runtimeMarkdownPath}");
+                        continue;
+                    }
+
+                    var markdownContent = await File.ReadAllTextAsync(runtimeMarkdownPath);
+                    var packages = GetPackagesFromMarkdown(runtimeMarkdownPath, markdownContent, warnings);
+                    foreach (var (packageName, packageVersion) in packages)
+                    {
+                        packagesAndVersions.Packages[packageName] = packageVersion;
+                    }
+                }
+            }
+        }
+
+        return (runtimePackages, warnings);
+    }
+
+    public static ImmutableArray<(string Name, SemVersion Version)> GetPackagesFromMarkdown(string markdownPath, string markdownContent, List<string> warnings)
+    {
+        var lines = markdownContent.Split("\n").Select(l => l.Trim()).ToArray();
+
+        // the markdown file contains a table that looks like this:
+        //   Package name | Version
+        //   :----------- | :------------------
+        //   Some.Package | 1.2.3
+        //   ...
+        // however there are some formatting issues with some elements that prevent markdown parsers from
+        // discovering it, so we fall back to manual parsing
+
+        var tableStartLine = -1;
+        for (int i = 0; i < lines.Length; i++)
+        {
+            if (Regex.IsMatch(lines[i], "Package name.*Version"))
```

Comment:
Nitpick: Do we need to protect against trailing whitespace?

Reply:
Shouldn't be a problem since I didn't anchor this with `^` and `$`.